### PR TITLE
New: Remove AlbumFolder, Support Nested Track Format

### PR DESCRIPTION
**Pull request by Qstick**
#### Database Migration
YES (42)

#### Description
Remove the AlbumFolder setting per artist and the AlbumFolder Naming Format
Allow Nesting in the Track and Artist Naming Format Fields

#### Todos
- [x] Tests
- [x] Add DB Migration to CONCAT Album Folder Format to track formats
- [ ] Documentation

### CODE REVIEW

File: src/NzbDrone.Core/Datastore/Migration/042_remove_album_folders.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(42)]
+    public class remove_album_folders : NzbDroneMigrationBase
+    {
+        protected override void MainDbUpgrade()
+        {
+            Delete.Column("AlbumFolder").FromTable("Artists");
+
+            Execute.Sql("UPDATE NamingConfig SET StandardTrackFormat = AlbumFolderFormat || '/' || StandardTrackFormat, MultiDiscTrackFormat = AlbumFolderFormat || '/' || MultiDiscTrackFormat");
```
Comment (Qstick) [inline src/NzbDrone.Core/Datastore/Migration/042_remove_album_folders.cs:13]: Should we be a bit smarter here and not update if user has disabled album folders for all artists?

Comment (ta264) [reply src/NzbDrone.Core/Datastore/Migration/042_remove_album_folders.cs:13]: Does it work on windows with / too?

**Review by ta264 — COMMENTED** [review]
I think this looks good. I was worrying about how the metadata service handles multi disk paths (so the album folder looks like `album/cd1/track`) but I think that was pretty broken before anyway :D

I think the fix is probably to get the common base dir for all tracks instead of looking at directory name of the first.

File: src/NzbDrone.Core/Datastore/Migration/042_remove_album_folders.cs (hunk 1 above)
Comment (Qstick) [reply src/NzbDrone.Core/Datastore/Migration/042_remove_album_folders.cs:13]: Yea, the name builder splits on either and recombines using path.combine

**Review by ta264 — APPROVED** [review]

**Comment by Qstick** [comment]
Should be good for a look

File: src/NzbDrone.Core/Organizer/NamingConfig.cs [hunk 2]
```diff
@@ -8,17 +8,15 @@ public class NamingConfig : ModelBase
         {
             RenameTracks = false,
             ReplaceIllegalCharacters = true,
-            StandardTrackFormat = "{Artist Name} - {Album Title} - {track:00} - {Track Title}",
+            StandardTrackFormat = "{Album Title} ({Release Year})/{Artist Name} - {Album Title} - {track:00} - {Track Title}",
             MultiDiscTrackFormat = "{Medium Format} {medium:00}/{Artist Name} - {Album Title} - {track:00} - {Track Title}",
```
[the remark this answers is not in the document]
Comment (Qstick) [reply src/NzbDrone.Core/Organizer/NamingConfig.cs:12]: Yep good call, thought I had done it

File: src/NzbDrone.Core/MediaFiles/TrackFileMovingService.cs [hunk 3]
```diff
@@ -132,13 +132,6 @@ private TrackFile TransferFile(TrackFile trackFile, Artist artist, List<Track> t
             try
             {
                 _mediaFileAttributeService.SetFolderLastWriteTime(artist.Path, trackFile.DateAdded);
-
-                if (artist.AlbumFolder)
-                {
-                    var albumFolder = Path.GetDirectoryName(destinationFilePath);
-
-                    _mediaFileAttributeService.SetFolderLastWriteTime(albumFolder, trackFile.DateAdded);
```
[the remark this answers is not in the document]
Comment (Qstick) [reply src/NzbDrone.Core/MediaFiles/TrackFileMovingService.cs:140]: We probably do need it at some point, I suspect it was already not working for medium sub folders etc...

**Pull request merged**